assets/src/edit-story/components/panels/layer/provider.js

```diff
@@ -26,20 +26,77 @@ import StoryPropTypes from '../../../types';
 import Context from './context';
 import useLayers from './useLayers';
 
+const SCROLL_DISTANCE = 5;
+
 function LayerProvider({ children }) {
   const layers = useLayers();
+  const [scrollRatio, setScrollRatio] = useState(null);
   const [isReordering, setIsReordering] = useState(false);
+  const [scrollDirection, setScrollDirection] = useState(0);
   const [currentSeparator, setCurrentSeparator] = useState(null);
+  const scrollTarget = useRef(null);
+
+  const setScrollTarget = useCallback(
+    (node) => (scrollTarget.current = node.parentElement),
+    []
+  );
+
+  const updateScrollRatio = useCallback(() => {
+    const node = scrollTarget.current;
+    if (!node) return;
+    const possibleScroll = node.scrollHeight - node.clientHeight;
+    if (possibleScroll <= 0) {
+      setScrollRatio(null);
+    } else {
+      setScrollRatio(node.scrollTop / possibleScroll);
+    }
+  }, []);
+
+  const stopScroll = useCallback(() => setScrollDirection(0), [
+    setScrollDirection,
+  ]);
+
+  const canScrollUp = isReordering && scrollRatio > 0;
+  const canScrollDown = isReordering && scrollRatio < 1;
+
+  useEffect(() => {
+    const isNotScrolling = scrollDirection === 0;
+    const isScrollingUp = canScrollUp && scrollDirection < 0;
+    const isScrollingDown = canScrollDown && scrollDirection > 0;
+    const isScrollingAnywhere = isScrollingUp || isScrollingDown;
+    if (isNotScrolling || !isScrollingAnywhere) {
+      return;
+    }
+
+    scrollTarget.current.scrollTop += scrollDirection * SCROLL_DISTANCE;
+    updateScrollRatio();
```

Comment:
This is the cheeky bit. This updates `scrollRatio` which causes the component to re-render - and including `scrollRatio` in the deps array then makes this effect re-run.

Reply:
Hm, it might be too cheeky. React (occasionally) complains that `Maximum update depth exceeded`. I'll rewrite to RAF...

Reply:
I agree - should definitely do a `requestAnimationFrame` and probably cancel it on cleanup and then we should also use `useBatchingCallback` for `updateScrollRatio`.

But my alternative suggestion is also this:
 1. `useEffect()` on `startScroll` and `scrollDirection`.
 2. The effect will start interval and cancel it in the cleanup function.
 3. On each interval callback: read current scroll position, calc `canScrollX`, and increment/decrement the scroll position.

This way, there's no re-rendering involved on each cycle. And such a scrolling implementation is more similar to the actual manual browser scroll. Another thing this could allow in the future: since the whole scrolling process will be captured within a single `useEffect/setInterval` we can play with the momentum. E.g. a longer scroll could pick up speed.